flink-core/src/main/java/org/apache/flink/core/io/SimpleVersionedSerializerTypeSerializerProxy.java

```diff
@@ -0,0 +1,163 @@
+		try {
+			return new SimpleVersionedSerializerTypeSerializerProxy<>(
+					InstantiationUtil.deserializeObject(
+							serializedSerializer,
+							serializerSupplier.getClass().getClassLoader()));
+		} catch (ClassNotFoundException | IOException e) {
+			throw new RuntimeException("Could not duplicate SimpleVersionedSerializer.", e);
+		}
+	}
+
+	@Override
+	public T createInstance() {
+		return null;
+	}
+
+	@Override
+	public T copy(T from) {
+		SimpleVersionedSerializer<T> serializer = getSerializer();
+		try {
+			byte[] serializedFrom = serializer.serialize(from);
+			return serializer.deserialize(
+					serializer.getVersion(),
+					serializedFrom);
+		} catch (IOException e) {
+			throw new RuntimeException("Could not copy element.", e);
+		}
+	}
+
+	@Override
+	public T copy(T from, T reuse) {
+		// the reuse is optional, we can just ignore it
+		return copy(from);
+	}
+
+	@Override
+	public int getLength() {
+		return -1;
+	}
+
+	@Override
+	public void serialize(T record, DataOutputView target) throws IOException {
+		SimpleVersionedSerializer<T> serializer = getSerializer();
+		target.writeInt(serializer.getVersion());
+		byte[] serializedRecord = serializer.serialize(record);
+		target.writeInt(serializedRecord.length);
+		target.write(serializedRecord);
+	}
+
+	@Override
+	public T deserialize(DataInputView source) throws IOException {
+		SimpleVersionedSerializer<T> serializer = getSerializer();
+		int version = source.readInt();
+		int length = source.readInt();
+		byte[] serializedRecord = new byte[length];
+		source.readFully(serializedRecord);
+		return serializer.deserialize(version, serializedRecord);
+	}
+
+	@Override
+	public T deserialize(T reuse, DataInputView source) throws IOException {
+		// the reuse is optional, we can just ignore it
+		return deserialize(source);
+	}
+
+	@Override
+	public void copy(DataInputView source, DataOutputView target) throws IOException {
+		target.writeInt(source.readInt()); // version
+		int length = source.readInt();
+		target.writeInt(length); // length
+		byte[] serializedRecord = new byte[length];
+		source.readFully(serializedRecord);
+		target.write(serializedRecord);
+	}
+
+	@Override
+	public boolean equals(Object other) {
+		return other instanceof SimpleVersionedSerializerTypeSerializerProxy
+				&& ((SimpleVersionedSerializerTypeSerializerProxy<?>) other).serializerSupplier
+				.get()
+				.equals(serializerSupplier.get());
+	}
+
+	@Override
+	public int hashCode() {
+		return serializerSupplier.get().hashCode();
+	}
+
+	@Override
+	public TypeSerializerSnapshot<T> snapshotConfiguration() {
+		throw new UnsupportedOperationException(
+				"SimpleVersionedSerializerWrapper is not meant to be used as a general TypeSerializer for state");
+	}
+
+	private SimpleVersionedSerializer<T> getSerializer() {
+		if (cachedSerializer != null) {
+			return cachedSerializer;
+		}
+		cachedSerializer = serializerSupplier.get();
+		return cachedSerializer;
+	}
+
+	@FunctionalInterface
```

Comment:
If you want, maybe move it to a `Utils` class as it may be useful in other places as well.